internal/tofu/context_plugins.go

```diff
@@ -68,72 +80,27 @@ func (cp *contextPlugins) NewProvisionerInstance(typ string) (provisioners.Inter
-	provider, err := cp.NewProviderInstance(addr)
-	if err != nil {
-		return schemas, fmt.Errorf("failed to instantiate provider %q to obtain schema: %w", addr, err)
-	}
-	defer provider.Close(ctx)
-
-	resp := provider.GetProviderSchema(ctx)
-	if resp.Diagnostics.HasErrors() {
-		return resp, fmt.Errorf("failed to retrieve schema from provider %q: %w", addr, resp.Diagnostics.Err())
-	}
-
-	if resp.Provider.Version < 0 {
-		// We're not using the version numbers here yet, but we'll check
-		// for validity anyway in case we start using them in future.
-		return resp, fmt.Errorf("provider %s has invalid negative schema version for its configuration blocks,which is a bug in the provider ", addr)
-	}
-
-	for t, r := range resp.ResourceTypes {
-		if err := r.Block.InternalValidate(); err != nil {
-			return resp, fmt.Errorf("provider %s has invalid schema for managed resource type %q, which is a bug in the provider: %w", addr, t, err)
-		}
-		if r.Version < 0 {
-			return resp, fmt.Errorf("provider %s has invalid negative schema version for managed resource type %q, which is a bug in the provider", addr, t)
-		}
-	}
-
-	for t, d := range resp.DataSources {
-		if err := d.Block.InternalValidate(); err != nil {
-			return resp, fmt.Errorf("provider %s has invalid schema for data resource type %q, which is a bug in the provider: %w", addr, t, err)
-		}
-		if d.Version < 0 {
-			// We're not using the version numbers here yet, but we'll check
-			// for validity anyway in case we start using them in future.
-			return resp, fmt.Errorf("provider %s has invalid negative schema version for data resource type %q, which is a bug in the provider", addr, t)
-		}
-	}
-
-	for t, d := range resp.EphemeralResources {
-		if err := d.Block.InternalValidate(); err != nil {
-			return resp, fmt.Errorf("provider %s has invalid schema for ephemeral resource type %q, which is a bug in the provider: %w", addr, t, err)
-		}
-		if d.Version < 0 {
-			// We're not using the version numbers here yet, but we'll check
-			// for validity anyway in case we start using them in future.
-			return resp, fmt.Errorf("provider %s has invalid negative schema version for ephemeral resource type %q, which is a bug in the provider", addr, t)
-		}
+	// Coarse lock only for ensuring that a valid entry exists
+	cp.providerSchemasLock.Lock()
+	entry, ok := cp.providerSchemas[addr]
+	if !ok {
+		entry = sync.OnceValues(func() (providers.ProviderSchema, error) {
+			log.Printf("[TRACE] tofu.contextPlugins: Initializing provider %q to read its schema", addr)
+			provider, err := cp.NewProviderInstance(addr)
+			if err != nil {
+				return providers.ProviderSchema{}, fmt.Errorf("failed to instantiate provider %q to obtain schema: %w", addr, err)
+			}
+			defer provider.Close(ctx)
+
+			schema := provider.GetProviderSchema(ctx)
+			return schema, schema.Validate(addr)
+		})
+		cp.providerSchemas[addr] = entry
```

Comment:
To make things clearer regarding the lock, I would extract this logic into another function and use defer in there, or just use an anonymous func to do so:
```go
func() {
	cp.providerSchemasLock.Lock()
 	defer cp.providerSchemasLock.Unlock()
    entry, ok := cp.providerSchemas[addr]
    if !ok {
        entry := ...
        cp.providerSchemas[addr] = entry
    }
}()
```

This is trying to also tackle the comment Martin left on this one. Obviously, the suggested code is kinda ugly so I would suggest to create a fully separate unexported method 🤷

Reply:
I've tinkered around with this a bit, and IMO it's a lot less readable.